File: src/functions/touchpad.py
import pygame
import time
import sys
from datetime import datetime

from src import app_state
from src.functions import dev_mode


def draw_text(lines, color=(255, 255, 255)):
    """Desenha texto centralizado na tela seguindo o padrão do main.py"""
    if app_state.SCREEN is None:
        return

    app_state.SCREEN.fill((0, 0, 0))
    y = app_state.HEIGHT // 3
    for text in lines:
        rendered = app_state.FONT.render(text, True, color)
        rect = rendered.get_rect(center=(app_state.WIDTH // 2, y))
        app_state.SCREEN.blit(rendered, rect)
        y += 50
    pygame.display.flip()


def touchpad_step():
    """
    Testa o touchpad/mouse em uma tela unica:
    - Arraste ate a caixa da esquerda e clique com o botao esquerdo
    - Arraste ate a caixa da direita e clique com o botao direito
    - Botao do meio (opcional)
    - Scroll para cima e para baixo
    """
    SCREEN = app_state.SCREEN
    WIDTH = app_state.WIDTH
    HEIGHT = app_state.HEIGHT
    FONT = app_state.FONT
    CLOCK = app_state.CLOCK
    
    
    drag_start = None
    scroll_up_count = 0
    scroll_down_count = 0

    # Log de testes realizados
    log_data = []

    completed = {
        "drag_left": False,
        "left_click": False,
        "drag_right": False,
        "right_click": False,
        "scroll_up": False,
        "scroll_down": False,
        "middle_click": False,
    }

    required_steps = [
        "drag_left",
        "left_click",
        "drag_right",
        "right_click",
        "scroll_up",
        "scroll_down",
    ]

    # Layout responsivo
    def clamp(value, min_value, max_value):
        return max(min_value, min(value, max_value))

    box_w = clamp(WIDTH // 3, 200, 420)
    box_h = clamp(HEIGHT // 4, 120, 260)
    box_y = HEIGHT // 2 - box_h // 2
    left_rect = pygame.Rect(WIDTH // 4 - box_w // 2, box_y, box_w, box_h)
    right_rect = pygame.Rect(WIDTH * 3 // 4 - box_w // 2, box_y, box_w, box_h)

    def mark_done(key, log_step):
        if not completed[key]:
            completed[key] = True
            log_data.append({"step": log_step, "time": str(datetime.now())})

    def all_required_done():
        return all(completed[key] for key in required_steps)

    def draw_ui():
        SCREEN.fill((0, 0, 0))

        # Titulos
        title = FONT.render("Teste de Touchpad", True, (255, 255, 255))
        SCREEN.blit(title, title.get_rect(center=(WIDTH // 2, clamp(HEIGHT // 8, 60, 140))))

        # Caixa esquerda (botao esquerdo)
        left_color = (0, 200, 0) if completed["drag_left"] and completed["left_click"] else (100, 100, 100)
        pygame.draw.rect(SCREEN, left_color, left_rect, 3)
        left_label = FONT.render("Clique ESQUERDO", True, (255, 255, 255))
        SCREEN.blit(left_label, left_label.get_rect(center=left_rect.center))

        # Caixa direita (botao direito)
        right_color = (0, 200, 0) if completed["drag_right"] and completed["right_click"] else (100, 100, 100)
        pygame.draw.rect(SCREEN, right_color, right_rect, 3)
        right_label = FONT.render("Clique DIREITO", True, (255, 255, 255))
        SCREEN.blit(right_label, right_label.get_rect(center=right_rect.center))

        # Checklist
        items = [
            ("Arraste ate a caixa da esquerda", completed["drag_left"], False),
            ("Clique ESQUERDO na caixa", completed["left_click"], False),
            ("Arraste ate a caixa da direita", completed["drag_right"], False),
            ("Clique DIREITO na caixa", completed["right_click"], False),
            (f"Scroll PARA CIMA (2x) {scroll_up_count}/2", completed["scroll_up"], False),
            (f"Scroll PARA BAIXO (2x) {scroll_down_count}/2", completed["scroll_down"], False),
            ("Botao do meio", completed["middle_click"], True),
        ]

        y = clamp(HEIGHT // 3, 180, HEIGHT // 2)
        for text, done, optional in items:
            prefix = "[OK] " if done else "[ ] "
            suffix = " (opcional)" if optional else ""
            color = (0, 255, 0) if done else (200, 200, 200)
            rendered = FONT.render(prefix + text + suffix, True, color)
            SCREEN.blit(rendered, rendered.get_rect(center=(WIDTH // 2, y)))
            y += clamp(HEIGHT // 30, 22, 36)

        # Aprovacao/Reprovacao
        approve_enabled = all_required_done()
        status_text = "Y = aprovar, N = reprovar" if approve_enabled else "Complete os passos obrigatorios"
        status_color = (0, 200, 0) if approve_enabled else (200, 200, 200)
        status_render = FONT.render(status_text, True, status_color)
        SCREEN.blit(status_render, status_render.get_rect(center=(WIDTH // 2, HEIGHT * 5 // 6)))

        pygame.display.flip()
    
    while True:
        for event in pygame.event.get():
            dev_mode.handle(event)
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_ESCAPE:
                    return log_data

                if event.key == pygame.K_y and all_required_done():
                    log_data.append({"step": "TOUCHPAD_APPROVED", "time": str(datetime.now())})
                    return log_data

                if event.key == pygame.K_n:
                    log_data.append({"step": "TOUCHPAD_REPROVED", "time": str(datetime.now())})
                    return log_data
            
            # Detecção de movimento do mouse
            elif event.type == pygame.MOUSEMOTION:
                if drag_start is None:
                    drag_start = event.pos
                else:
                    dx = event.pos[0] - drag_start[0]
                    dy = event.pos[1] - drag_start[1]
                    drag_distance = (dx * dx + dy * dy) ** 0.5

                    if drag_distance >= 100 and left_rect.collidepoint(event.pos):
                        mark_done("drag_left", "TOUCHPAD_DRAG_TO_LEFT_BOX")
                        drag_start = None

                    elif drag_distance >= 100 and right_rect.collidepoint(event.pos):
                        mark_done("drag_right", "TOUCHPAD_DRAG_TO_RIGHT_BOX")
                        drag_start = None
            
            # Cliques do mouse
            elif event.type == pygame.MOUSEBUTTONDOWN:
                # Botao esquerdo
                if event.button == 1 and left_rect.collidepoint(event.pos):
                    mark_done("left_click", "TOUCHPAD_LEFT_CLICK")

                # Botao direito
                elif event.button == 3 and right_rect.collidepoint(event.pos):
                    mark_done("right_click", "TOUCHPAD_RIGHT_CLICK")

                # Botao do meio (opcional)
                elif event.button == 2:
                    mark_done("middle_click", "TOUCHPAD_MIDDLE_CLICK")

                # Scroll para cima
                elif event.button == 4:
                    if not completed["scroll_up"]:
                        scroll_up_count += 1
                        if scroll_up_count >= 2:
                            mark_done("scroll_up", "TOUCHPAD_SCROLL_UP")

                # Scroll para baixo
                elif event.button == 5:
                    if not completed["scroll_down"]:
                        scroll_down_count += 1
                        if scroll_down_count >= 2:
                            mark_done("scroll_down", "TOUCHPAD_SCROLL_DOWN")

        draw_ui()
        CLOCK.tick(10)



if __name__ == "__main__":
    import sys
    import os
    
    # Adiciona o diretório raiz ao path para permitir imports absolutos
    root_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../..'))
    sys.path.insert(0, root_dir)
    
    # Simular ambiente de app_state para teste standalone
    pygame.init()
    app_state.WIDTH, app_state.HEIGHT = 1920, 1080
    app_state.SCREEN = pygame.display.set_mode((app_state.WIDTH, app_state.HEIGHT))
    pygame.display.set_caption("Teste de Touchpad")
    app_state.FONT = pygame.font.SysFont("Arial", 20)
    app_state.CLOCK = pygame.time.Clock()
    
    print("Iniciando teste de touchpad...")
    results = touchpad_step()
    
    print("\n" + "="*50)
    print("RESULTADOS DO TESTE DE TOUCHPAD")
    print("="*50)
    
    for log in results:
        print(f"{log['step']}: {log['time']}")
    
    print("="*50)
    print(f"\n✓ {len(results)} testes concluídos com sucesso!")
    
    pygame.quit()
    sys.exit(0)

File: crates/revy-check/src/steps/touchpad.rs
//! Touchpad / mouse test (mirrors `touchpad.touchpad_step`): drag to each box +
//! click, scroll up/down twice, optional middle click, then Y/N.

use macroquad::prelude::*;

use crate::app::App;
use crate::ui::draw;

fn clamp(v: f32, lo: f32, hi: f32) -> f32 {
    v.max(lo).min(hi)
}

#[derive(Default)]
struct Done {
    drag_left: bool,
    left_click: bool,
    drag_right: bool,
    right_click: bool,
    scroll_up: bool,
    scroll_down: bool,
    middle_click: bool,
}

impl Done {
    fn required_done(&self) -> bool {
        self.drag_left
            && self.left_click
            && self.drag_right
            && self.right_click
            && self.scroll_up
            && self.scroll_down
    }
}

pub async fn run(app: &mut App) {
    let mut done = Done::default();
    let mut drag_start: Option<Vec2> = None;
    let mut scroll_up_count = 0;
    let mut scroll_down_count = 0;

    loop {
        app.tick_global().await;

        let w = screen_width();
        let h = screen_height();
        let box_w = clamp(w / 3.0, 200.0, 420.0);
        let box_h = clamp(h / 4.0, 120.0, 260.0);
        let box_y = h / 2.0 - box_h / 2.0;
        let left_rect = Rect::new(w / 4.0 - box_w / 2.0, box_y, box_w, box_h);
        let right_rect = Rect::new(w * 3.0 / 4.0 - box_w / 2.0, box_y, box_w, box_h);

        let (mx, my) = mouse_position();
        let mouse = vec2(mx, my);

        // Drag detection: accumulate distance from a fixed start until a box is reached.
        match drag_start {
            None => drag_start = Some(mouse),
            Some(start) => {
                let dist = mouse.distance(start);
                if dist >= 100.0 && left_rect.contains(mouse) {
                    if !done.drag_left {
                        done.drag_left = true;
                        app.log.add_now("TOUCHPAD_DRAG_TO_LEFT_BOX", "APROVADO");
                    }
                    drag_start = None;
                } else if dist >= 100.0 && right_rect.contains(mouse) {
                    if !done.drag_right {
                        done.drag_right = true;
                        app.log.add_now("TOUCHPAD_DRAG_TO_RIGHT_BOX", "APROVADO");
                    }
                    drag_start = None;
                }
            }
        }

        // Clicks.
        if is_mouse_button_pressed(MouseButton::Left) && left_rect.contains(mouse) && !done.left_click {
            done.left_click = true;
            app.log.add_now("TOUCHPAD_LEFT_CLICK", "APROVADO");
        }
        if is_mouse_button_pressed(MouseButton::Right) && right_rect.contains(mouse) && !done.right_click {
            done.right_click = true;
            app.log.add_now("TOUCHPAD_RIGHT_CLICK", "APROVADO");
        }
        if is_mouse_button_pressed(MouseButton::Middle) && !done.middle_click {
            done.middle_click = true;
            app.log.add_now("TOUCHPAD_MIDDLE_CLICK", "APROVADO");
        }

        // Scroll (need two notches each way).
        let (_, wheel_y) = mouse_wheel();
        if wheel_y > 0.0 && !done.scroll_up {
            scroll_up_count += 1;
            if scroll_up_count >= 2 {
                done.scroll_up = true;
                app.log.add_now("TOUCHPAD_SCROLL_UP", "APROVADO");
            }
        } else if wheel_y < 0.0 && !done.scroll_down {
            scroll_down_count += 1;
            if scroll_down_count >= 2 {
                done.scroll_down = true;
                app.log.add_now("TOUCHPAD_SCROLL_DOWN", "APROVADO");
            }
        }

        // Approve / reject.
        if is_key_pressed(KeyCode::Y) && done.required_done() {
            app.log.add_now("TOUCHPAD_APPROVED", "APROVADO");
            return;
        }
        if is_key_pressed(KeyCode::N) {
            app.log.add_now("TOUCHPAD_REPROVED", "REPROVADO");
            return;
        }

        draw_ui(app, &done, left_rect, right_rect, scroll_up_count, scroll_down_count);
        next_frame().await;
    }
}

fn draw_ui(
    app: &App,
    done: &Done,
    left_rect: Rect,
    right_rect: Rect,
    scroll_up_count: i32,
    scroll_down_count: i32,
) {
    let w = screen_width();
    let h = screen_height();
    let font = app.font();

    clear_background(BLACK);
    draw::text_centered("Teste de Touchpad", w / 2.0, clamp(h / 8.0, 60.0, 140.0), 22, WHITE, font);

    let left_color = if done.drag_left && done.left_click { draw::rgb(0, 200, 0) } else { draw::rgb(100, 100, 100) };
    draw_rectangle_lines(left_rect.x, left_rect.y, left_rect.w, left_rect.h, 3.0, left_color);
    draw::text_centered("Clique ESQUERDO", left_rect.x + left_rect.w / 2.0, left_rect.y + left_rect.h / 2.0, 22, WHITE, font);

    let right_color = if done.drag_right && done.right_click { draw::rgb(0, 200, 0) } else { draw::rgb(100, 100, 100) };
    draw_rectangle_lines(right_rect.x, right_rect.y, right_rect.w, right_rect.h, 3.0, right_color);
    draw::text_centered("Clique DIREITO", right_rect.x + right_rect.w / 2.0, right_rect.y + right_rect.h / 2.0, 22, WHITE, font);

    let items = [
        ("Arraste ate a caixa da esquerda".to_string(), done.drag_left, false),
        ("Clique ESQUERDO na caixa".to_string(), done.left_click, false),
        ("Arraste ate a caixa da direita".to_string(), done.drag_right, false),
        ("Clique DIREITO na caixa".to_string(), done.right_click, false),
        (format!("Scroll PARA CIMA (2x) {}/2", scroll_up_count), done.scroll_up, false),
        (format!("Scroll PARA BAIXO (2x) {}/2", scroll_down_count), done.scroll_down, false),
        ("Botao do meio".to_string(), done.middle_click, true),
    ];

    let mut y = clamp(h / 3.0, 180.0, h / 2.0);
    for (text, ok, optional) in &items {
        let prefix = if *ok { "[OK] " } else { "[ ] " };
        let suffix = if *optional { " (opcional)" } else { "" };
        let color = if *ok { draw::rgb(0, 255, 0) } else { draw::rgb(200, 200, 200) };
        draw::text_centered(&format!("{prefix}{text}{suffix}"), w / 2.0, y, 22, color, font);
        y += clamp(h / 30.0, 22.0, 36.0);
    }

    let approve = done.required_done();
    let status = if approve { "Y = aprovar, N = reprovar" } else { "Complete os passos obrigatorios" };
    let status_color = if approve { draw::rgb(0, 200, 0) } else { draw::rgb(200, 200, 200) };
    draw::text_centered(status, w / 2.0, h * 5.0 / 6.0, 22, status_color, font);
}
